### PR TITLE
[device][adc] implement adc_get_vref

### PR DESCRIPTION
**Pull request by byte-me-stan**
add stm32_adc_get_vref

## 拉取/合并请求描述：(PR description)

[
这段方括号里的内容是您**必须填写并替换掉**的，否则PR不可能被合并。**方括号外面的内容不需要修改，但请仔细阅读。**
The content in this square bracket must be filled in and replaced, otherwise, PR can not be merged. The contents outside square brackets need not be changed, but please read them carefully.

请在这里填写您的PR描述，可以包括以下之一的内容：为什么提交这份PR；解决的问题是什么，你的解决方案是什么；
Please fill in your PR description here, which can include one of the following items: why to submit this PR; what is the problem solved and what is your solution;

并确认并列出已经在什么情况或板卡上进行了测试。
And confirm in which case or board has been tested.
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [ ] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [ ] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [ ] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [ ] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [ ] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/contribution_guide/coding_style_en.txt)

### CODE REVIEW

File: bsp/stm32/libraries/HAL_Drivers/drv_adc.c [hunk 1]
```diff
@@ -169,7 +170,11 @@ static rt_uint32_t stm32_adc_get_channel(rt_uint32_t channel)
 
     return stm32_channel;
 }
-
+static rt_uint32_t stm32_adc_get_vref (struct rt_adc_device *device)
+{
+    RT_ASSERT(device);
+    return 3300;
+}
```
Comment (mysterywolf) [inline bsp/stm32/libraries/HAL_Drivers/drv_adc.c:177]: There should add a blank line between next function

File: components/drivers/include/drivers/adc.h [hunk 2]
```diff
@@ -34,6 +35,7 @@ typedef enum
     RT_ADC_CMD_ENABLE = RT_DEVICE_CTRL_BASE(ADC) + 1,
     RT_ADC_CMD_DISABLE = RT_DEVICE_CTRL_BASE(ADC) + 2,
     RT_ADC_CMD_GET_RESOLUTION = RT_DEVICE_CTRL_BASE(ADC) + 3,
+    RT_ADC_CMD_GET_VREF = RT_DEVICE_CTRL_BASE(ADC) + 4,
```
Comment (mysterywolf) [inline components/drivers/include/drivers/adc.h:38]:
```suggestion
    RT_ADC_CMD_GET_VREF = RT_DEVICE_CTRL_BASE(ADC) + 4, /* get reference voltage */
```

File: components/drivers/misc/adc.c [hunk 3]
```diff
@@ -64,6 +63,10 @@ static rt_err_t _adc_control(rt_device_t dev, int cmd, void *args)
             result = RT_EOK;
         }
     }
+    else if (cmd == RT_ADC_CMD_GET_VREF && adc->ops->get_resolution)
+    {
+        result = adc->ops->get_vref(adc);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:68]: we cannot directly return the voltage value, the result is rt_err_t type.
```suggestion
    else if (cmd == RT_ADC_CMD_GET_VREF && adc->ops->get_vref && args)
    {
        rt_int16_t value;

        value= adc->ops->get_vref(adc);
        *((rt_int16_t*)args) = value;
        if (value == 0)
        {
            result = -RT_ENOSYS;
        }
        else
        {
            result = RT_EOK;
        }
```

File: components/drivers/misc/adc.c [hunk 4]
```diff
@@ -156,17 +159,20 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
-   rt_uint32_t value = 0, voltage = 0;
+    rt_uint32_t value = 0, voltage = 0, vref = 0;
 
     RT_ASSERT(dev);
 
     /*read the value and convert to voltage*/
     if (dev->ops->get_resolution != RT_NULL && dev->ops->convert != RT_NULL)
     {
+        /*get reference voltage*/
+        vref = _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, RT_NULL );
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:170]: Use adc->ops->get_vref to instead.

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/drivers/misc/adc.c (hunk 3 above)
Comment (mysterywolf) [reply components/drivers/misc/adc.c:68]: Please help me to add arg check in adc->ops->get_resolution branch

File: components/drivers/misc/adc.c [hunk 5]
```diff
@@ -156,17 +159,20 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
+
         /*get the convert bits*/
         rt_uint8_t resolution = dev->ops->get_resolution(dev);
         dev->ops->convert(dev, channel, &value);
-        voltage = value * REFER_VOLTAGE / (1 << resolution);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:169]: Need to change:
1. use _adc_control to instead dev->ops->get_resolution and dev->ops->convert, which can give you a clear error code.
2. need to fill in the `arg` parameter

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/drivers/misc/adc.c (hunk 3 above)
Comment (mysterywolf) [reply components/drivers/misc/adc.c:68]: I think rt_int16_t , not rt_uint16_t or rt_uint32_t

File: components/drivers/include/drivers/adc.h (hunk 2 above)
Comment (mysterywolf) [reply components/drivers/include/drivers/adc.h:38]: /* get the resolution in bits */

File: components/drivers/misc/adc.c (hunk 3 above)
Comment (mysterywolf) [reply components/drivers/misc/adc.c:68]: all of the value of reference voltage changes from  rt_uint32_t to rt_int16_t

File: components/drivers/misc/adc.c (hunk 4 above)
Comment (mysterywolf) [inline components/drivers/misc/adc.c:170]: if (_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
{
    return 0;
}

File: components/drivers/misc/adc.c [hunk 6]
```diff
@@ -156,17 +159,20 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
     RT_ASSERT(dev);
 
     /*read the value and convert to voltage*/
     if (dev->ops->get_resolution != RT_NULL && dev->ops->convert != RT_NULL)
     {
+        /*get reference voltage*/
+        vref = _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, RT_NULL );
+
         /*get the convert bits*/
         rt_uint8_t resolution = dev->ops->get_resolution(dev);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:173]: if (_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_RESOLUTION , &xxxx) != RT_EOK)
{
return 0;
}

File: components/drivers/misc/adc.c [hunk 7]
```diff
@@ -154,19 +162,25 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
+    rt_int16_t vref = 0, voltage = 0;
+    rt_uint8_t resolution;
 
     RT_ASSERT(dev);
 
+    /*get the convert bits*/
+    _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_RESOLUTION, &resolution);
+
+    /*get the reference voltage*/
+    _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref);
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:177]: Please check the error code

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/drivers/misc/adc.c [hunk 8]
```diff
@@ -154,28 +162,42 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
 {
-   rt_uint32_t value = 0, voltage = 0;
+    rt_uint32_t value = 0;
+    rt_int16_t vref = 0, voltage = 0;
+    rt_uint8_t resolution = 0;
 
     RT_ASSERT(dev);
 
+    /*get the convert bits*/
+    if(_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_RESOLUTION, &resolution) != RT_EOK)
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:174]: if blank space (  xxxx)

File: components/drivers/misc/adc.c [hunk 9]
```diff
@@ -154,28 +162,42 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
+    {
+        goto _voltage_exit;
+    }
+
+    /*get the reference voltage*/
+   if( _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:180]:
```suggestion
   if ( _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
```

File: components/drivers/misc/adc.c [hunk 10]
```diff
@@ -154,28 +162,42 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
+   {
+       goto _voltage_exit;
+   }
+
     /*read the value and convert to voltage*/
-    if (dev->ops->get_resolution != RT_NULL && dev->ops->convert != RT_NULL)
+    if(resolution != RT_NULL && vref != RT_NULL)
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:186]: we don't need that if branch any more

**Review by mysterywolf — CHANGES_REQUESTED** [review]
Please carefully crossing check with the file that I sent to you.

File: components/drivers/misc/adc.c (hunk 9 above)
Comment (mysterywolf) [reply components/drivers/misc/adc.c:180]: Please care about the format. 4 spaces rather than 3 spaces. if + a blank space + (

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/drivers/misc/adc.c [hunk 11]
```diff
@@ -154,28 +162,42 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
-        rt_uint8_t resolution = dev->ops->get_resolution(dev);
         dev->ops->convert(dev, channel, &value);
-        voltage = value * REFER_VOLTAGE / (1 << resolution);
+        voltage = value * vref / (1 << resolution);
     }
 
+_voltage_exit:
     return voltage;
 }
 #ifdef RT_USING_FINSH
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:196]:
```suggestion
```

File: components/drivers/misc/adc.c [hunk 12]
```diff
@@ -154,28 +162,42 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
-        /*get the convert bits*/
-        rt_uint8_t resolution = dev->ops->get_resolution(dev);
         dev->ops->convert(dev, channel, &value);
-        voltage = value * REFER_VOLTAGE / (1 << resolution);
+        voltage = value * vref / (1 << resolution);
     }
 
+_voltage_exit:
     return voltage;
 }
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:194]:
```suggestion
}

```

**Review by mysterywolf — CHANGES_REQUESTED** [review]
Please help me to add and delete these two lines

File: components/drivers/misc/adc.c [hunk 13]
```diff
@@ -154,28 +162,40 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
-        rt_uint8_t resolution = dev->ops->get_resolution(dev);
-        dev->ops->convert(dev, channel, &value);
-        voltage = value * REFER_VOLTAGE / (1 << resolution);
+        goto _voltage_exit;
     }
 
+    /*get the reference voltage*/
+    if( _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
+    {
+       goto _voltage_exit;
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:182]: 3 spaces

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/drivers/misc/adc.c [hunk 14]
```diff
@@ -154,28 +162,40 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
-   rt_uint32_t value = 0, voltage = 0;
+    rt_uint32_t value = 0;
+    rt_int16_t vref = 0, voltage = 0;
+    rt_uint8_t resolution = 0;
 
     RT_ASSERT(dev);
 
-    /*read the value and convert to voltage*/
-    if (dev->ops->get_resolution != RT_NULL && dev->ops->convert != RT_NULL)
+    /*get the convert bits*/
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:173]:
```suggestion
    /*get the resolution in bits*/
```

File: components/drivers/misc/adc.c [hunk 15]
```diff
@@ -154,28 +162,40 @@ rt_err_t rt_adc_disable(rt_adc_device_t dev, rt_uint32_t channel)
...
     {
-        /*get the convert bits*/
-        rt_uint8_t resolution = dev->ops->get_resolution(dev);
-        dev->ops->convert(dev, channel, &value);
-        voltage = value * REFER_VOLTAGE / (1 << resolution);
+        goto _voltage_exit;
     }
 
+    /*get the reference voltage*/
+    if( _adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
```
Comment (mysterywolf) [inline components/drivers/misc/adc.c:180]:
```suggestion
    if(_adc_control((rt_device_t) dev, RT_ADC_CMD_GET_VREF, &vref) != RT_EOK)
```

**Review by mysterywolf — CHANGES_REQUESTED** [review]

**Review by mysterywolf — APPROVED** [review]

**Comment by mysterywolf** [comment]
请Squash合并

**Review by Guozhanxin — APPROVED** [review]

**Pull request merged**